### PR TITLE
GH-33734: [Go] make compatible with grpc < 1.45

### PR DESCRIPTION
**Pull request by chrisirhc**
### Rationale for this change

Define interface from grpc/reflection rather than importing. The interface is marked as experiment, so defining it here is also a step toward improving stability of the go arrow library.

### What changes are included in this PR?

* As above, inlining the interface

### Are these changes tested?

I believe the the change is already tested via:
https://github.com/apache/arrow/blob/c8d6110a26c41966e539e9fa2f5cb8c31dc2f0fe/go/arrow/flight/flight_test.go#L325-L326
And also via confirming via reading https://github.com/grpc/grpc-go/blob/4e4d8288ff4fb99150a526165c767e6fac3bd5ec/Documentation/server-reflection-tutorial.md . However, I have not done any manual testing beyond this.

### Are there any user-facing changes?
No.

### Feedback requested
I'd pasted the code verbatim from: https://github.com/grpc/grpc-go/blob/4c776ec01572d55249df309251900554b46adb41/reflection/serverreflection.go#L69-L83

Should I have pasted only what's necessary and link to that code block instead?

### CODE REVIEW

File: go/arrow/flight/server.go [hunk 1]
```diff
@@ -64,6 +63,24 @@ func RegisterFlightServiceServer(s grpc.ServiceRegistrar, srv FlightServer) {
...
+// for a custom implementation to return zero values for the
+// grpc.ServiceInfo values in the map.
+//
+// Experimental
+//
+// Notice: This type is EXPERIMENTAL and may be changed or removed in a
+// later release.
+type serviceInfoProvider interface {
+	GetServiceInfo() map[string]grpc.ServiceInfo
+}
```
Comment (zeroshade) [inline go/arrow/flight/server.go:82]: Can you add a small comment about the fact that we're inlining this in order to provide compatibility with grpc < 1.45?

**Review by zeroshade — CHANGES_REQUESTED** [review]
LGTM just one nitpick

File: go/arrow/flight/server.go (hunk 1 above)
Comment (chrisirhc) [reply go/arrow/flight/server.go:82]: Yep sure thing. Any thoughts about the comment copied verbatim from https://github.com/grpc/grpc-go/blob/4c776ec01572d55249df309251900554b46adb41/reflection/serverreflection.go#L69-L83 ?

**Comment by chrisirhc** [comment]
Addressed comment.

File: go/arrow/flight/server.go (hunk 1 above)
Comment (zeroshade) [reply go/arrow/flight/server.go]: I think it's fine copying the comment verbatim for now.

**Review by zeroshade — APPROVED** [review]
Thanks much @chrisirhc ! once the CI passes I'll merge this

**Pull request merged**